### PR TITLE
Identifier scanning

### PR DESCRIPTION
**Pull request by KFlash**
Need at least a second or third view on this code.  Identifier scanning is one of the most important jobs for the lexer and has to be fast.  

- Any overheads?
- Optimization tweak ideas?
- Ideas for simplification?

6k tests running locale with 91% coverage. 4040 tests and now 37% coverage in CLI. Same files.

### CODE REVIEW

**Comment by g-plane** [comment]
I haven't seen any notable changes. Lines just be joined, which is not readable.

**Comment by KFlash** [comment]
@g-plane Can you make this readable? I have to do some RL stuff. And then merge if no other feedback?

File: packages/cherow/src/lexer/identifier.ts [hunk 1]
```diff
@@ -19,20 +19,19 @@ import {
...
- * @param first codepoint
  */
 export function scanIdentifier(parser: Parser): Token {
   const { index } = parser;
   let first = parser.source.charCodeAt(parser.index);
   // Hot path - fast scanning for identifiers and non-escaped keywords
   while (isAsciiCodePoint(first)) {
-      parser.index++;
-      parser.column++;
+      parser.index++; parser.column++;
```
Comment (g-plane) [inline packages/cherow/src/lexer/identifier.ts]: Split them into two lines. There are some code like this below.

**Comment by KFlash** [comment]
@g-plane Fixed it. In this PR support for `💩` is missing. Has to be in a separate PR, but I need to start focus on the parser tiself now. Do you want to add in support for it?

File: packages/cherow/src/lexer/identifier.ts [hunk 2]
```diff
@@ -19,49 +19,48 @@ import {
...
```
Comment (g-plane) [inline packages/cherow/src/lexer/identifier.ts]: Missing here.

File: packages/cherow/src/lexer/identifier.ts [hunk 3]
```diff
@@ -97,39 +96,37 @@ function parseIdentifierSuffix(parser: Parser): Token {
 export function scanIdentifierUnicodeEscape(parser: Parser): number {
   parser.index++;
   parser.column++;
-  if (parser.source.charCodeAt(parser.index) !== Chars.LowerU) report(parser, Errors.Unexpected);
-  parser.index++;
-  parser.column++;
+  if (parser.source.charCodeAt(parser.index) !== Chars.LowerU) report(parser, Errors.InvalidUnicodeEscape);
+  parser.index++; parser.column++;
```
Comment (g-plane) [inline packages/cherow/src/lexer/identifier.ts]: Missing here.

File: packages/cherow/src/lexer/identifier.ts [hunk 4]
```diff
@@ -97,39 +96,37 @@ function parseIdentifierSuffix(parser: Parser): Token {
...
       let value = 0;
       let digit = toHex(parser.source.charCodeAt(parser.index));
       if (digit < 0) return -1;
       while (digit >= 0) {
           value = (value << 4) | digit;
-          if (value > Chars.NonBMPMax) report(parser, Errors.Unexpected);
-          parser.index++;
-          parser.column++;
+          if (value > Chars.NonBMPMax) report(parser, Errors.UndefinedUnicodeCodePoint);
+          parser.index++; parser.column++;
```
Comment (g-plane) [inline packages/cherow/src/lexer/identifier.ts]: Missing here.

**Comment by g-plane** [comment]
@KFlash Emoji as identifier?

**Comment by KFlash** [comment]
What's left to fix is to recognize identifier characters from the Supplementary Multilingual Plane (SMP)

**Comment by g-plane** [comment]
So is it hard to fix?

**Comment by KFlash** [comment]
I doubt that. But if you can't, just merge this and open a issue ticket for it.

**Pull request merged**